src/snowflake/snowpark/_internal/open_telemetry.py

```diff
@@ -0,0 +1,79 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+import functools
+import inspect
+import os.path
+from contextlib import contextmanager
+from logging import getLogger
+from typing import Dict, List, Tuple
+
+logger = getLogger(__name__)
+target_class = ["dataframe.py", "dataframe_writer.py"]
+# this parameter make sure no error when open telemetry is not installed
+open_telemetry_found = True
+try:
+    from opentelemetry import trace
+
+except ImportError:
+    open_telemetry_found = False
+
+if open_telemetry_found:
+    tracer = trace.get_tracer("snow.snowpark.dataframe")
+
+
+@contextmanager
+def open_telemetry_context_manager(func, dataframe):
+    try:
+        # trace when required package is installed
+        if open_telemetry_found:
+            name = func.__qualname__
+            with tracer.start_as_current_span(name) as cur_span:
+                try:
+                    if cur_span.is_recording():
+                        # store execution location in span
+                        filename, lineno = context_manager_code_location(inspect.stack(), func)
+                        cur_span.set_attribute("code.filepath", f"{filename}")
+                        cur_span.set_attribute("code.lineno", lineno)
+                        # stored method chain
+                        method_chain = build_method_chain(dataframe._plan.api_calls, name)
+                        cur_span.set_attribute("method.chain", method_chain)
+                except Exception as e:
```

Comment:
When do we expect an exception? I ask because I'm thinking of using logger.warning in the next line.